platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreator.java

```diff
@@ -0,0 +1,355 @@
+    private final Time time;
+    private final Random random;
+    private final Signer signer;
+    private final AddressBook addressBook;
+    private final long selfId;
+    private final TipsetBuilder tipsetBuilder;
+    private final TipsetScoreCalculator tipsetScoreCalculator;
+    private final ChildlessEventTracker childlessEventTracker;
+    private final TransactionSupplier transactionSupplier;
+    private final SoftwareVersion softwareVersion;
+
+    // TODO add bully score and encapsulate metrics in a helper class
+    private static final RunningAverageMetric.Config TIPSET_SCORE_CONFIG = new RunningAverageMetric.Config(
+                    "platform", "tipsetScore")
+            .withDescription("The score, based on tipset advancements, of each new event created by this "
+                    + "node. A score of 0.0 means the event did not advance consensus at all, while a score "
+                    + "of 1.0 means that the event advanced consensus as much as a single event can.");
+    private final RunningAverageMetric tipsetScoreMetric;
+
+    /**
+     * The last event created by this node. TODO we need to load this if we don't start from genesis
+     */
+    private EventFingerprint lastSelfEvent;
+
+    /**
+     * Create a new tipset event creator.
+     *
+     * @param platformContext     the platform context
+     * @param cryptography        the cryptography instance
+     * @param time                provides wall clock time
+     * @param random              a source of randomness, does not need to be cryptographically secure
+     * @param signer              used for signing things with this node's private key
+     * @param addressBook         the current address book
+     * @param selfId              this node's ID
+     * @param softwareVersion     the current software version of the application
+     * @param transactionSupplier provides transactions to be included in new events
+     */
+    public TipsetEventCreator(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final Cryptography cryptography,
+            @NonNull final Time time,
+            @NonNull final Random random,
+            @NonNull final Signer signer,
+            @NonNull final AddressBook addressBook,
+            final long selfId,
+            @NonNull final SoftwareVersion softwareVersion,
+            @NonNull final TransactionSupplier transactionSupplier) {
+
+        this.cryptography = Objects.requireNonNull(cryptography);
+        this.time = Objects.requireNonNull(time);
+        this.random = Objects.requireNonNull(random);
+        this.signer = Objects.requireNonNull(signer);
+        this.addressBook = Objects.requireNonNull(addressBook);
+        this.selfId = selfId;
+        this.transactionSupplier = Objects.requireNonNull(transactionSupplier);
+        this.softwareVersion = Objects.requireNonNull(softwareVersion);
+
+        // TODO reduce indirection in the lambdas
```

Comment:
By which comment I think you mean that the address book changes infrequently so you might just cache the  index-to-weight map?  Then update it when the address book sends you a "changed" event?  (Unless the address book is actually constant during a given run in which case it's even easier?)